sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/SymmetricHashJoinStateManager.scala

```diff
@@ -187,17 +217,17 @@ class SymmetricHashJoinStateManager(
 
       // Find the next value satisfying the condition, updating `currentKey` and `numValues` if
       // needed. Returns null when no value can be found.
-      private def findNextValueForIndex(): UnsafeRow = {
+      private def findNextValueForIndex(): (UnsafeRow, Boolean) = {
         // Loop across all values for the current key, and then all other keys, until we find a
         // value satisfying the removal condition.
         def hasMoreValuesForCurrentKey = currentKey != null && index < numValues
         def hasMoreKeys = allKeyToNumValues.hasNext
         while (hasMoreValuesForCurrentKey || hasMoreKeys) {
           if (hasMoreValuesForCurrentKey) {
             // First search the values for the current key.
-            val currentValue = keyWithIndexToValue.get(currentKey, index)
+            val (currentValue, matched) = keyWithIndexToValue.getWithMatched(currentKey, index)
```

Comment:
`getWithMatched` can return `null`; is it guaranteed that it won't at this point? Otherwise you'll get an error here.

Maybe `KeyWithIndexToValueRowConverterFormatV2` should return `(null, false)` instead of  `null`? It would at least be closer to the semantics of the previous code.

And now that I noticed it, I kinda dislike `getWithMatched` for the same reason I disliked the other type names.

Is the extra overhead of `valueRowConverter.convertValue` allocating a new tuple, and this code "unapplying" that tuple, something that we need to worry about? I'm thinking that maybe if you return a proper type and take the same reuse approach as is done with `KeyToValuePair`, that would be better?

(But don't know if that can be done safely here.)

Reply:
Actually, it's more than just the allocation; every time you call `KeyWithIndexToValueRowConverterFormatV2.convertValue` you will not only allocate that new tuple (and unapply it, depending on the call site), but you'll also be applying a projection to the `UnsafeRow` from the state store.

Is that overhead acceptable here?

Reply:
> I'm thinking that maybe if you return a proper type and take the same reuse approach as is done with KeyToValuePair, that would be better?

The optimization around reusing instance is based on the fact that setter provides the iterator. Actually current optimization was really odd for me - after some digging I understood the intention but without knowing such detail, it's super easy to mess up. Iterating two `getAll()` results concurrently would provide interesting result.

Getting an element with applying reuse approach would be really confusing. Maybe it can be acceptable if we invert the approach and let caller provides the instance to fill (say, "out parameter") - caller is responsible to not let the instance be messed up, but easier to do that given they have context, especially if caller calls in a loop. I'll check if it could be done safely.

Regarding projection, we've done similar things in my previous PR (removing duplicated columns from key part to reduce state store size - the patch applied projection to key row - performance didn't hurt much (more clearly, the size of delta significantly matters) - https://github.com/apache/spark/pull/21733#issuecomment-411207042)

Reply:
I'd wait for hearing voice on out parameter, as it complicates the logic (all methods in call hierarchy should support out parameter to achieve object reuse). I've changed tuple to case class to deal with null issue.